File: D_Counting_Factorizations.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <map>
#include <cmath>
using namespace std;

const int MOD = 998244353;

// Compute the prime factorization of a number
unordered_map<int, int> prime_factorization(int n) {
    unordered_map<int, int> factors;
    for (int i = 2; i <= sqrt(n); i++) {
        while (n % i == 0) {
            factors[i]++;
            n /= i;
        }
    }
    if (n > 1) {
        factors[n]++;
    }
    return factors;
}

// Compute the number of positive integers that have a given multiset of prime factors
int count_factors(vector<int>& a) {
    int n = a.size() / 2;
    map<vector<pair<int, int>>, int> freq;
    for (int i = 0; i < (1 << n); i++) {
        vector<pair<int, int>> factors;
        for (int j = 0; j < n; j++) {
            int count = (i >> j) & 1 ? a[n+j] : a[j];
            if (count > 0) {
                factors.push_back({a[j], count});
            }
        }
        freq[factors]++;
    }
    int ans = 0;
    for (auto p : freq) {
        vector<pair<int, int>> factors = p.first;
        int count = p.second;
        int prod = 1;
        for (auto f : factors) {
            prod = (prod * (f.second + 1)) % MOD;
        }
        ans = (ans + (prod * count) % MOD) % MOD;
    }
    return ans;
}

int main() {
    int n;
    cin >> n;
    vector<int> a(2*n);
    for (int i = 0; i < 2*n; i++) {
        cin >> a[i];
    }
    cout << count_factors(a) << endl;
    return 0;
}

File: D_Counting_Factorizations.py
#include <bits/stdc++.h>
using namespace std;

const int MOD = 998244353;

int count_factors(vector<int>& a) {
    int n = a.size();
    vector<vector<int>> dp(1001, vector<int>(n+1));
    dp[2][0] = 1;
    for (int i = 1; i <= n; i++) {
        if (a[i-1] == 2) {
            dp[2][i] = dp[2][i-1];
        }
        for (int j = 3; j <= 1000; j++) {
            dp[j][i] = dp[j][i-1];
            if (a[i-1] % j == 0) {
                dp[j][i] = (dp[j][i] + dp[j-1][i]) % MOD;
            }
        }
    }
    int ans = 0;
    for (int j = 2; j <= 1000; j++) {
        ans = (ans + dp[j][n]) % MOD;
    }
    return ans;
}

int main() {
    int n;
    cin >> n;
    vector<int> a(2*n);
    for (int i = 0; i < 2*n; i++) {
        cin >> a[i];
    }
    cout << count_factors(a) << endl;
    return 0;
}
